pallets/capacity/src/types.rs

```diff
@@ -223,3 +246,300 @@ pub fn unlock_chunks_from_vec<T: Config>(chunks: &Vec<(u32, u32)>) -> UnlockChun
+	///
+	/// Note there is no sense of what the current era is; subsequent calls could return a different result
+	/// if 'reward_era' is the current era and there has been a boost or unstake.
+	pub(crate) fn get_amount_staked_for_era(&self, reward_era: &RewardEra) -> BalanceOf<T> {
+		// this gives an ordered-by-key Iterator
+		let mut bmap_iter = self.0.iter();
+		let mut eligible_amount: BalanceOf<T> = Zero::zero();
+		while let Some((era, balance)) = bmap_iter.next() {
+			if era.eq(reward_era) {
+				return *balance;
+			}
+			// there was a boost or unstake in this era.
+			else if era.gt(reward_era) {
+				return eligible_amount;
+			} // eligible_amount has been staked through reward_era
+			eligible_amount = *balance;
+		}
+		eligible_amount
+	}
+
+	/// Returns the number of history items
+	pub fn count(&self) -> usize {
+		self.0.len()
+	}
+
+	fn remove_oldest_entry_if_full(&mut self) {
+		if self.is_full() {
+			// compiler errors with unwrap
+			if let Some((earliest_key, _earliest_val)) = self.0.first_key_value() {
+				self.0.remove(&earliest_key.clone());
+			}
+		}
+	}
+
+	fn get_last_staking_amount(&self) -> BalanceOf<T> {
+		// compiler errors with unwrap
+		if let Some((_last_key, last_value)) = self.0.last_key_value() {
+			return *last_value;
+		};
+		Zero::zero()
+	}
+
+	fn is_full(&self) -> bool {
+		self.count().eq(&(T::ProviderBoostHistoryLimit::get() as usize))
+	}
+}
+
+/// Struct with utilities for storing and updating unlock chunks
+#[derive(Debug, TypeInfo, PartialEqNoBound, EqNoBound, Clone, Decode, Encode, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct RetargetInfo<T: Config> {
+	/// How many times the account has retargeted this RewardEra
+	pub retarget_count: u32,
+	/// The last RewardEra they retargeted
+	pub last_retarget_at: RewardEra,
+	_marker: PhantomData<T>,
+}
+
+impl<T: Config> Default for RetargetInfo<T> {
+	fn default() -> Self {
+		Self { retarget_count: 0u32, last_retarget_at: Zero::zero(), _marker: Default::default() }
+	}
+}
+
+impl<T: Config> RetargetInfo<T> {
+	/// Constructor
+	pub fn new(retarget_count: u32, last_retarget_at: RewardEra) -> Self {
+		Self { retarget_count, last_retarget_at, _marker: Default::default() }
+	}
+	/// Increment retarget count and return Some() or
+	/// If there are too many, return None
+	pub fn update(&mut self, current_era: RewardEra) -> Option<()> {
+		let max_retargets = T::MaxRetargetsPerRewardEra::get();
+		if self.retarget_count.ge(&max_retargets) && self.last_retarget_at.eq(&current_era) {
+			return None;
+		}
+		if self.last_retarget_at.lt(&current_era) {
+			self.last_retarget_at = current_era;
+			self.retarget_count = 1;
+		} else {
+			self.retarget_count = self.retarget_count.saturating_add(1u32);
+		}
+		Some(())
+	}
+}
+
+/// A trait that provides the Economic Model for Provider Boosting.
+pub trait ProviderBoostRewardsProvider<T: Config> {
```

Comment:
It seems that Hash, Balance, and AccountId are not being used in this trait. Instead, we are using a type alias for Balance. We should consider removing these unused associated types to simplify the trait, or alternatively, ensure Self::Balance is consistently used if Balance is required.

example
```
	fn era_staking_reward(
		era_amount_staked: Self::Balance,
		era_total_staked: Self::Balance,
		era_reward_pool_size: Self::Balance,
	) -> Self::Balance;
```

Reply:
They were being used by a function that was removed, so yea we should get rid of them.

Reply:
done